rbd/encryption.go

```diff
@@ -0,0 +1,113 @@
+// +build !octopus,!nautilus
+
+package rbd
+
+// #cgo LDFLAGS: -lrbd
+// /* force XSI-complaint strerror_r() */
+// #define _POSIX_C_SOURCE 200112L
+// #undef _GNU_SOURCE
+// #include <errno.h>
+// #include <stdlib.h>
+// #include <rados/librados.h>
+// #include <rbd/librbd.h>
+import "C"
+
+import (
+	"unsafe"
+)
+
+// EncryptionFormat encryption format type
+type EncryptionFormat C.uint
+
+// Possible values for EncryptionFormat:
+// ENCRYPTION_FORMAT_LUKS1: LUKS v1
+// ENCRYPTION_FORMAT_LUKS2: LUKS v2
+const (
+	ENCRYPTION_FORMAT_LUKS1 = EncryptionFormat(C.RBD_ENCRYPTION_FORMAT_LUKS1)
+	ENCRYPTION_FORMAT_LUKS2 = EncryptionFormat(C.RBD_ENCRYPTION_FORMAT_LUKS2)
+)
+
+// EncryptionAlgorithm encryption algorithm
+type EncryptionAlgorithm C.uint
+
+// Possible values for EncryptionAlgorithm:
+// ENCRYPTION_ALGORITHM_AES128: AES 128bits
+// ENCRYPTION_ALGORITHM_AES256: AES 256bits
+const (
+	ENCRYPTION_ALGORITHM_AES128 = EncryptionAlgorithm(C.RBD_ENCRYPTION_ALGORITHM_AES128)
+	ENCRYPTION_ALGORITHM_AES256 = EncryptionAlgorithm(C.RBD_ENCRYPTION_ALGORITHM_AES256)
+)
+
+// EncryptionFormatOptsLUKS1 and EncryptionFormatOptsLUKS2 are identical
+// structures at the moment, just as they are in the librbd api.
+// The purpose behind creating different identical structures, is to facilitate
+// future modifications of one of the formats, while maintaining backwards
+// compatibility with the other.
+
+// EncryptionFormatOptsLUKS1 options required for LUKS v1
+type EncryptionFormatOptsLUKS1 struct {
+	Alg        EncryptionAlgorithm
+	Passphrase []byte
+}
+
+// EncryptionFormatOptsLUKS2 options required for LUKS v2
+type EncryptionFormatOptsLUKS2 struct {
+	Alg        EncryptionAlgorithm
+	Passphrase []byte
+}
+
+// EncryptionOptions interface is used to encapsulate the different encryption
+// formats options and enable converting them from go to C structures.
+type EncryptionOptions interface {
+	cephEncryptionOptions() (C.rbd_encryption_options_t, C.size_t)
+}
+
+func (opts EncryptionFormatOptsLUKS1) cephEncryptionOptions() (C.rbd_encryption_options_t, C.size_t) {
+	var cOptsSize C.size_t
+	var cOptsPtr C.rbd_encryption_options_t
+	var cOpts C.rbd_encryption_luks1_format_options_t
+	cOpts.alg = C.rbd_encryption_algorithm_t(opts.Alg)
+	cOpts.passphrase = (*C.char)(C.CBytes(opts.Passphrase))
```

Comment:
CBytes copies the content into C memory and [must be freed](https://pkg.go.dev/cmd/cgo#hdr-Go_references_to_C). We'll want to call that free after you call rbd_encryption_format

Reply:
Which is not so easy, because `cOpts` is casted to a `rbd_encryption_options_t`, which makes passphrase unreachable.

Reply:
Good catch.  I added a C.free call in freeEncryptionOptions as follows.  Looks legit?  

```
   type EncryptionOptions interface {
      allocateEncryptionOptions() (C.rbd_encryption_options_t, C.size_t)
      freeEncryptionOptions(cOptsPtr C.rbd_encryption_options_t)
   }
   
   func (opts EncryptionFormatOptsLUKS1) freeEncryptionOptions(cOptsPtr C.rbd_encryption_options_t) {
      C.free(unsafe.Pointer(((*C.rbd_encryption_luks1_format_options_t)(cOptsPtr)).passphrase))
   }
   
   cOptsPtr, cOptsSize := opts.allocateEncryptionOptions()
   ret := C.rbd_encryption_format(
      image.image,
      C.rbd_encryption_format_t(format),
      cOptsPtr,
      cOptsSize)
   opts.freeEncryptionOptions(cOptsPtr)
```

Reply:
Well, it "works", but it definitely has a [code smell](https://en.m.wikipedia.org/wiki/Code_smell). 😉 Jumping through hoops like that is almost always a sign of issues with the code structure. To do it "right" you would need another interface for keeping the original struct, and it's getting more and more bloated. For me it's a clear case of over-engineering. I totally understand the emotional attachment to code we put effort in already, but I would try to overcome that feeling and go with a leaner solution. 😉 Also it would be closer to the original C call, what we usually try to do.

Reply:
I think this approach is OK. I like only having one argument as an interface type. It seems more "self documenting" to me. The casting is a tad bit ugly but we have to deal with the fact we're dealing with C's form of "polymorphism" somewhere in this code anyway :-)
I'll let you folks debate more if you like but since all the ugly bits are part of the private calls I am fine with it. I'll try do do another pass tomorrow but I just wanted to throw my 2 cents in on this aspect early.

Reply:
Fortunately there's no need for the ugly cast. How about this leaner interface then:

```go
type EncryptionOptions interface {
	cephEncryptionOptions() (
		C.rbd_encryption_format_t, C.rbd_encryption_options_t, C.size_t, func())
}

func (image *Image) EncryptionFormat(opts EncryptionOptions) error {
	if image.image == nil {
		return ErrImageNotOpen
	}

	cFormat, cOptsPtr, cOptsSize, cOptsFree := opts.cephEncryptionOptions()
	defer cOptsFree()

	ret := C.rbd_encryption_format(
		image.image,
		cFormat,
		cOptsPtr,
		cOptsSize)
	return getError(ret)
} 
```

The implementation of cephEncryptionOptions() can look like:
```go
func (opts EncryptionFormatOptsLUKS1) cephEncryptionOptions() (
	C.rbd_encryption_format_t, C.rbd_encryption_options_t, C.size_t, func()) {
...
	cOpts.passphrase = (*C.char)(C.CBytes(opts.Passphrase))
	cOptsFree := func() { C.free(unsafe.Pointer(cOpts.passphrase)) }
...
	return C.rbd_encryption_format_t(EncryptionFormatLUKS1), cOptsPtr, cOptsSize, cOptsFree
}
```

Reply:
Coming from C background, I kind of prefer the cast :P  But I can see the advantage in your approach @ansiwen.  I implemented it and will push the update momentarily.

Reply:
Yeah, we all carry our baggage from the past. 🤷‍♂️😜